plugin/src/automation/lift/liftrules.ml

```diff
@@ -167,33 +171,90 @@ let is_packed_constr c env sigma trm =
+        sigma, None
+  | UnpackSigma ->
+     if (get_lifting c).is_fwd then
        if is_packed c trm then
-          let sigma_right, args_opt = type_is_from c env trm sigma in
-          if Option.has_some args_opt then
-            let sigma = sigma_right in
-            let constr = constrs.(0) in
-            let pms = Option.get args_opt in
-            let args = pair_projections_eta_rec_n trm (arity constr - List.length pms) in
-            sigma, Some (0, List.append pms args)
-          else
-            sigma, None
+         (* we treat any existential of the right type as a constructor *)
+         let sigma_right, args_opt = type_is_from c env trm sigma in
+         if Option.has_some args_opt then
+           let typ_args = Option.get args_opt in
+           let sigma, b_sig_eq =
+             let b_sig_eq_typ = mkAppl (fst (get_types c), typ_args) in
+             Util.on_snd dest_sigT (reduce_term env sigma_right b_sig_eq_typ)
+           in
+           let s, h_eq =
+             let trm_ex = dest_existT trm in
+             trm_ex.index, trm_ex.unpacked
+           in
+           let i_b, b =
+             if is_or_applies existT s then
+               let index_ex = dest_existT s in
+               index_ex.index, index_ex.unpacked
+             else
+               projections (dest_sigT b_sig_eq.index_type) s
+           in sigma, Some (0, List.append typ_args [i_b; b; h_eq])
+         else
+           sigma, None
+       else
+         sigma, None
+     else (* <-- TODO correct or not? needed? also, this is slow *)
+       let sigma_right, args_opt = type_is_from c env trm sigma in
+       if Option.has_some args_opt then
+         let typ_args = Option.get args_opt in
+         if (is_or_applies eq_rect trm) then (* TODO better as elim rule. maybe want whole thing as elim rule and no constrs? *)
+           let sigma, trm = expand_eta env sigma_right trm in
+           let eq_args = Array.of_list (unfold_args trm) in
+           let i_b = eq_args.(1) in
+           let b = eq_args.(3) in
+           let h_eq = eq_args.(5) in
+           if is_or_applies eq_refl h_eq then
+             sigma, None
+           else
+             sigma, Some (0, List.append typ_args [i_b; b; h_eq])
+         else
+           if isApp trm && isConstruct (first_fun trm) then (* TODO blehhh *)
+             (* TODO get correct args; then return actual constr; then refold correctly *)
+             let i_b = last typ_args in
+             let sigma, i_b_typ = reduce_type env sigma_right i_b in
+             let b = trm in
+             let h_eq = apply_eq_refl { typ = i_b_typ; trm = i_b } in
+             sigma, Some (0, List.append typ_args [i_b; b; h_eq])
+           else
+             sigma, None
        else
          sigma, None
 
 (* Premises for LIFT-PACK *)
 let is_pack c env sigma trm =
   let l = get_lifting c in
   let right_type trm = type_is_from c env trm sigma in
-  if l.is_fwd then
+  if l.is_fwd || l.orn.kind = UnpackSigma then
```

Comment:
Rename to "is_identity," and explain more and clean a bit. Consolidate constr rule stuff here, and maybe_repack as well